docs: add mkdocs

GabDug opened the pull request:
Early draft to fix #401 

Preview at https://gabdug.github.io/django-components/dev/

To manage the documentation environment, [hatch](https://hatch.pypa.io/latest/) with hatch-mkdocs and hatch-pip-compile has been used.

`pipx install hatch`
`hatch run docs:serve`

## Features

- [X] Build the documentation using mkdocs and mkdocs-material
  - Lots of extensions and plugins, rich rendering is possible
- [X] Version the documentation with mike
  - Push on a `gh-pages` branch
- [X] Generate the API reference with mkdocstrings
  - We can add references to our API, Django docs and Python docs
  - We generate an `objects.inv` file that can be used by others to link to our API

## Tasks

- [X] CI to generate on release (tag and `latest` versions)
- [X] CI to generate the `dev` version
- [X] CI to generate the documentation on PRs/branches
- [ ] API Reference should not have empty pages
  - Either add some documentation (docstring on module, class, function or variable) or remove some pages in `scripts/gen_ref_nav.py`
- [ ] Review and revamp the documentation
- [ ] Cleanup the plugins and extensions
- [ ] Cleanup the configuration
- [ ] Review header levels in the documentation
- [ ] Documentation on hatch
- [ ] Update CI to remove testing mode (all the XXX)
  - [ ] CI should be disabled on forks
- [ ] Fix the Github README
  - Should present quickly and redirect to the documentation for more information
  - Links should be to latest/dev doc
- [ ] Fix the Pypi README
  - Pypi README can be generated on the fly (with `hatch-fancy-pypi-readme` for instance)
  - Can have more information than the Github README, but should still redirect to the documentation for more
  - Link should be to the released version (not latest/dev)

## Nice to have

- [ ] Logo and Favicon?
- [ ] Choose a color scheme, or create a custom one?
- [ ] Inline examples are included automatically from tests or the example project
- [ ] Add mkdocsstrings references where appropriate (in code and in docs)
- [ ] Contributors page?
- [ ] Proper "added in version x.x" macro?

EmilStenstrom commented:
Wow, this is sooo much better! I'm **really** happy with this change :D

General quick thoughts:
* It feels a bit much to have both a top header with navigation and a left menu with navigation. Only use the left nav and skip the top nav?

Start page:
* Badge: I see that 3.6 and 3.7 are in the list?
* Features: Add "feature" that talks about rendering components are views for use with JS libraries like htmx
* Features: Single file components *or multi file components* as clear options
* Summary: We should have a simple example with slots too.

EmilStenstrom commented:
@GabDug What do we do about this PR. Close it? It's starting to be so far behind the original docs I'm not sure how we'll be able to merge it later. 

Maybe a simpler way to accomplish this is to build out the structure first, merge that, and then move the content in a separate PR?

GabDug commented:
Hey @EmilStenstrom!

Sorry for letting this PR rot. I've rebased from the main branch, and removed all documentation content, only including the current README. The documentation data migration can thus happen in another PR. 

Let me know that you think.

EmilStenstrom commented:
@JuroOravec I like this. Would you mind reviewing this?

JuroOravec commented:
Looks amazing! There's a lot of things to potentially touch on, now I understand why there's so many tasks listed 😄 But they can be all handled in due time.

JuroOravec commented on docs/CHANGELOG.md:
```diff
@@ -0,0 +1,13 @@
+
+{!
+    include-markdown "../README.md"
+    start="## Release notes"
```
I assume that this setup is temporary, so we can integrate with the current README, and then down the line we'd break the README down into separate files?

EmilStenstrom replied:
Yes, each section can probably be a separate file I think.

JuroOravec commented on scripts/gen_ref_nav.py:
```diff
@@ -0,0 +1,47 @@
+root = Path(__file__).parent.parent
+src = root / "src"
+
+for path in sorted(src.rglob("*.py")):
```
Down the line, my preference would be to document only the public API:
- `src/django_components/__init__.py`
- `src/django_components/templatetags/component_tags.py`
- `src/django_components/middleware.py`

JuroOravec commented on src/django_components/component_registry.py:
```diff
@@ -42,18 +42,19 @@ def clear(self) -> None:
 
-# This variable represents the global component registry
 registry: ComponentRegistry = ComponentRegistry()
+"""Global component registry."""
```
Note that there's updated docstring on the master branch.

JuroOravec commented on src/django_components/component_registry.py:
````diff
@@ -42,18 +42,19 @@ def clear(self) -> None:
-    @register("my_component")
-    class MyComponent(component.Component):
-        ...
+        ```python
````
Later, we will need to go over all the docstrings and update them to use the Google dostrings style as set here https://github.com/EmilStenstrom/django-components/pull/420/files#diff-98d0f806abc9af24e6a7c545d3d77e8f9ad57643e27211d7a7b896113e420ed2R184 (or set it to different style).

Till now, when I was writing docstrings, I formatted it so that it would look good for me on VSCode.

JuroOravec commented on pyproject.toml:
```diff
@@ -97,3 +97,37 @@ disallow_untyped_defs = true
     "tests"
 ]
+
+[tool.hatch.env]
```
I haven't heard of hatch before. But in [the docs](https://hatch.pypa.io/latest/why/) they position themselves as replacement for setuptools and tox. So I reckon that, down the line, we could migrate the hatch?

EmilStenstrom commented:
@GabDug You've got an approval! Mind rebasing on master so we can get this merged?! Thank you again! :)

EmilStenstrom commented:
I managed to rebase this from master, resolving all 13 merge conflicts, and now it's in. Thanks @GabDug!

GabDug commented:
@EmilStenstrom I'm glad to see it merged! Apologies for not being more active recently—I hope it wasn't too frustrating. Thanks for handling it!
